### PR TITLE
added some more standard JS global objects

### PR DESCRIPTION
**Pull request by jneuendorf**

### CODE REVIEW

File: heterarchy.litcoffee [hunk 1]
```diff
@@ -61,7 +61,7 @@ behaves like a class that would be have such a hierarchy.
         if isEqual linearization, hierarchy next
             next
         else
-            class result extends generate tail linearization
+            class Result extends generate tail linearization
```
Comment (jneuendorf) [inline heterarchy.litcoffee:63]: satisfy CoffeeLint

File: heterarchy.litcoffee [hunk 2]
```diff
@@ -91,7 +91,7 @@ it in the later.
     reparent = (oldklass, newklass, value) ->
         if value not instanceof Function
             value
-        else if value == oldklass::constructor and inherited(oldklass) == Object
+        else if value is oldklass::constructor and inherited(oldklass) is Object
```
Comment (jneuendorf) [inline heterarchy.litcoffee:93]: satisfy CoffeeLint

File: heterarchy.litcoffee [hunk 3]
```diff
@@ -152,8 +152,50 @@ mult-inherited classes:
...
+            SyntaxError
+            TypeError
+            URIError
         ]
 
+        if cls in standardClasses
+            true
+        else
+            # according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects
+            nonStandardClasses = [
```
Comment (jneuendorf) [inline heterarchy.litcoffee]: also support some of the newer classes

**Comment by arximboldi** [comment]
Thanks a lot for this contribution @jneuendorf, highly appreciated !

I will merge this later this evening. I have a couple of nitpicks:
- Commits should be in separate logical steps.This way it is easier to understand the code by using `git blame`. This means, I would sepparate ff790dd in two commits _"Make CoffeeLint happy"_ and _"Prevent adding an mro member to newer standard classes"_.
- Commits should pass tests. This way it is easier to find bugs with `git bisect`. This means squashing b7375a4 into the commit that broke the test.

A more technical nitpick: with your new list of classes we are now doing linear search (with the most common case being having to iterate over the whole list) over a non-trivial amount of elements. I might try to fix this later mysef, but it is no big deal since this code is not the most critical code path.

Anyways, I understand that this is your first PR to this repo so I will just merge your changes as-is.

Also, just out of curiosity: are you using the library in production or some other project? What is your use-case?

Once again, thanks!

**Comment by jneuendorf** [comment]
Hey, I've never created a pull request for a repo with travis + coverage. Therefore, I wasn't aware of those checks and I'll try to fix them. ;)
Also, I would like to squash the commits and improve what you've mentioned (O(n) for checking the classes)

I actually do use your heterarchy library in production (at work). We are building a single page web tool that stores company data (in the back end) and displays them in various form (in the front end - d3).
One example of where we use multi inheritance is the drawing graphs: there are grouped bar, stacked bar, grouped column, and stacked column charts, we started refactoring those chart classes so there are 4 parent classes: stacked chart, grouped chart, bar chart and column chart, and the according sub classes.
First I used RingJS (http://ringjs.neoname.eu/) but class properties and methods are not supported so I started rewriting the whole thing in CoffeeScript but as it turned out there were some bugs.
And then I found your lib in a comment in the coffee-script repo...so thank you for that comment. :D

So...does heterarchy actually support multi inheritance of class properties and methods? If not, I could propably just iterate the current MRO and take the first occurence. But in that case there must be static super method attached to each class to also support static MROs...

**Comment by jneuendorf** [comment]
How can I find out what parts of the code are uncovered?

**Comment by arximboldi** [comment]
You can check the per-line coverage here: https://coveralls.io/builds/7653995/source?filename=heterarchy.litcoffee

I believe that your change decreased coverage because you are not inheriting from any special class, so the case for special classes is actually never hit. One way to make your new code be checked would be to add a test that inherits from a special class available in node (e.g. `Map`) and then check whether the `__mro__` member has been added.

**W.r.t. to your last optimization**, I find it problematic that you add a `__isNative__` property to the fundamental class. Note that the whole reason that function exists at all is because I did not want to pollute system classes with custom attributes like `__mro__`, so adding an `__isNative__` attribute defeats its purpose.

The way I was thinking about optimizing it was more like computing a dictionary that maps auto-stringified classes to class objects. Then we would just check if the class is there. We could do this both for the common classes and the special classes, thus simplifying the code (and removing the need for a new test to fix coverage). Something like (I did not actually test this code and I haven't done Coffee for a few months, just for illustration):

``` coffeescript
javaScriptClassNames = [ "Object",  "Function" , ..., "Map", "Promise", ... ]
javaScriptClasses = _.reduce javaScriptClassNames, (classes, name) -> 
        classes[global[name]] = global[name]
        classes
  , {}
isJavaScriptClass = (cls) -> javaScriptClasses[cls] is cls
```

I'm glad that you are finding the library useful for work. A visual library is a good example of something that often benefits from multiple inheritance and mixins.

**W.r.t. to property support**, if you mean properties defined with `Objet.defineProperty`, there is no special support for it, but I'm not sure what exactly you mean. Do you want to call `super` in the getter/setter? Note that this is something that does not work in other languages neither (I'm thinking of python). The workaround in Python is to delegate to a method in the getter/setter. I think that in Coffee something like this should work (again, I did not actually try):

``` coffee
Function::property = (prop, desc) ->
    Object.defineProperty @prototype, prop, desc

class Person
    constructor: (@_name) ->
    @property 'name',
        get: -> do @_getName

class SeriousWoman extends Person
    _getName: -> "#{Ms. @super()}"

w = new SeriousWoman "Alice"
w.name is "Ms. Alice"
```

Cheers!

**Comment by jneuendorf** [comment]
Thanks for pointing out how to get an coverage overview. Is there a way to also see something like that on the command line?

I adjusted the `isJavaScriptClass` function as you suggested (which I find a very clever way to bypass the need to check potentially undefined variables with `typeof`!). That led to 100% coverage also.

**Concerning `static` support:** I haven't really though about the consequences yet but as described in https://github.com/nicolas-van/ring.js/issues/13 it would be nice to have something like this:

``` coffeescript
class A
    @prop: 42
    @classMethod: ->
        @prop

class B extends A
    @classMethod: ->
        return @super() + 2

B.classMethod().should.equal 44
# of course it should also work for multi inheritance
```

**Comment by arximboldi** [comment]
Oh I see what you mean. I think that should work for most cases. The only thing to be careful with is mutating the class level `prop` (mutating is the problem, but overriding in derived classes is ok). Because the way `multi` simulates MRO-based inheritance on top of JavaScript, it has to copy the classes that have to be put in the middle of a linearization with a tail different than its own. This effectively freezes the value of the class property, since changes to it are not visible through some derived classes. While this is ok for most sensible usages of class properties, it is something worth noting.

There is in fact a test documenting this limitation:
https://github.com/arximboldi/heterarchy/blob/master/test/heterarchy.spec.coffee#L198

**Comment by arximboldi** [comment]
Btw, thanks a lot for your changes and following up with my suggestions. Great work! I will merge now.

**Comment by jneuendorf** [comment]
Thank you, too, for being patient with me ;) I learned a lot! :) 👍

**Pull request merged**